netty/src/main/java/io/grpc/netty/WriteQueue.java

```diff
@@ -118,20 +128,35 @@ ChannelFuture enqueue(Object command, ChannelPromise promise, boolean flush) {
    * called in the event loop
    */
   private void flush() {
+    assert channel.eventLoop().inEventLoop();
+
     try {
       boolean flushed = false;
+      // We can't just call flush after having completely drained the queue, as new objects might
+      // be added to the queue continuously and so it may never get empty. If we never flushed in
+      // that case we would be guaranteed to OOM.
+      // However, we also don't want to flush too often as flushing invokes expensive system
+      // calls and we want to feed Netty with enough data so as to maximize the data written per
+      // system call.
+      int writesBeforeFlush = min(queue.size(), MAX_WRITES_BEFORE_FLUSH);
+      // Always dequeue in chunks, in order to not acquire the queue's lock too often.
       while (queue.drainTo(writeChunk, DEQUE_CHUNK_SIZE) > 0) {
-        while (writeChunk.size() > 0) {
-          writeChunk.poll().run();
+        writesBeforeFlush -= writeChunk.size();
+        while (!writeChunk.isEmpty()) {
+          QueuedCommand cmd = writeChunk.poll();
+          channel.write(cmd, cmd.promise());
+        }
+        if (writesBeforeFlush <= 0) {
+          writesBeforeFlush = min(queue.size(), MAX_WRITES_BEFORE_FLUSH);
```

Comment:
@ejona86 excellent point. yeah this code got more complicated than need be. I initially just had queue.size() and then figured it would be a good idea to also cap and added this code. I ll just use `MAX_WRITES_BEFORE_FLUSH`. That way the `flushed` parameter will make more sense again too.